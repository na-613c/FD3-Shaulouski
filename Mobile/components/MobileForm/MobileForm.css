.MobileForm {
    width: 400px;
    position: relative;
}


.MobileForm span {
    margin-right: 20px;
}

.MobileForm input {
    width: 200px;
    position: absolute;
    right: 0;
}