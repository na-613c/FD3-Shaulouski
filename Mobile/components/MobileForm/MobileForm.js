import React from 'react';
import PropTypes from 'prop-types';
import { mobileEvents } from '../../events';


import('./MobileForm.css');



class MobileForm extends React.PureComponent {

    static propTypes = {
        client: PropTypes.shape({
            id: PropTypes.number.isRequired,
            FIO: PropTypes.shape({
                fam: PropTypes.string.isRequired,
                im: PropTypes.string.isRequired,
                otch: PropTypes.string.isRequired,
            }),
            balance: PropTypes.number.isRequired,
        }),
        isEditMode: PropTypes.bool.isRequired,
    };


    constructor(props) {
        super(props);
        this.famEl = React.createRef();
        this.imEl = React.createRef();
        this.otchEl = React.createRef();
        this.balanceEl = React.createRef();
    }


    save = (EO) => {
        const elem = {
            id: this.props.client.id,
            FIO: {
                fam: this.famEl.current ? this.famEl.current.value : this.props.client.FIO.fam,
                im: this.imEl.current ? this.imEl.current.value : this.props.client.FIO.im,
                otch: this.otchEl.current ? this.otchEl.current.value : this.props.client.FIO.otch,
            },
            balance: +this.balanceEl.current ? +this.balanceEl.current.value : this.props.client.balance,
        }

        mobileEvents.emit('ESaveClient', elem);
    }

    create = () => {
        mobileEvents.emit('ECreate');
    }

    cancel = () => {
        mobileEvents.emit('ECancel');
    }


    render() {

        console.log('MobileForm render')

        if (this.props.isEditMode) {
            return (
                <div className='MobileForm' >
                    <div>
                        <span>Фамилия</span>
                        <input ref={this.famEl} type="text" defaultValue={this.props.client.FIO.fam} />
                    </div>
                    <div>
                        <span>Имя</span>
                        <input ref={this.imEl} type="text" defaultValue={this.props.client.FIO.im} />
                    </div>
                    <div>
                        <span>Отчество</span>
                        <input ref={this.otchEl} type="text" defaultValue={this.props.client.FIO.otch} />
                    </div>
                    <div>
                        <span>Баланс</span>
                        <input ref={this.balanceEl} type="number" defaultValue={this.props.client.balance} />
                    </div>
                    <button className='MobileForm-Save' onClick={this.save}>Сохранить</button>
                    <button className='MobileForm-Cancel' onClick={this.cancel}>Отмена</button>
                </div>
            )
        } else {
            return (
                <button className='MobileForm-Create' onClick={this.create}>
                    Создать
                </button>
            )
        }
    }
}

export default MobileForm;
