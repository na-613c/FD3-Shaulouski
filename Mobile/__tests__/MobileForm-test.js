import React from 'react';
import renderer from 'react-test-renderer';

import MobileForm from '../components/MobileForm/MobileForm';

test('работа MobileForm', () => {

    const client = { id: 101, FIO: { fam: "Иванов", im: "Иван", otch: "Иванович" }, balance: 200 };

    const componentActive = renderer.create(
        <MobileForm
            client={client}
            isEditMode={true}
        />
    );

    const componentPassive = renderer.create(
        <MobileForm
            client={client}
            isEditMode={false}
        />
    );

    // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
    let componentTreeActive = componentActive.toJSON();
    expect(componentTreeActive).toMatchSnapshot();

    let componentTreePassive = componentPassive.toJSON();
    expect(componentTreePassive).toMatchSnapshot();

    // найдём в вёрстке компонента саму кнопку
    const buttonSave = componentActive.root.findByProps({ className: "MobileForm-Save" })
    const buttonCancel = componentActive.root.findByProps({ className: "MobileForm-Cancel" })

    const buttonCreate = componentPassive.root.findByProps({ className: "MobileForm-Create" })

    // и "нажмём" на неё
    buttonSave.props.onClick();
    buttonCancel.props.onClick();

    buttonCreate.props.onClick();

    // получаем уже изменённый снэпшот
    componentTreeActive = componentActive.toJSON();
    expect(componentTreeActive).toMatchSnapshot();

    componentTreePassive = componentPassive.toJSON();
    expect(componentTreePassive).toMatchSnapshot();
    // "нажмём" кнопку ещё раз
    buttonSave.props.onClick();
    buttonCancel.props.onClick();

    buttonCreate.props.onClick();

    // и получаем окончательный снэпшот
    componentTreeActive = componentActive.toJSON();
    expect(componentTreeActive).toMatchSnapshot();

    componentTreePassive = componentPassive.toJSON();
    expect(componentTreePassive).toMatchSnapshot();
    /*
    // можно эмулировать события, передавая в качестве объекта события то что нам нужно:
    wrapper.find('select').simulate('change', {
        target: { value: "hello" },
    });
    */

});