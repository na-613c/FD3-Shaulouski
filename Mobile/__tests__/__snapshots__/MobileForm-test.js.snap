// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа MobileForm 1`] = `
<div
  className="MobileForm"
>
  <div>
    <span>
      Фамилия
    </span>
    <input
      defaultValue="Иванов"
      type="text"
    />
  </div>
  <div>
    <span>
      Имя
    </span>
    <input
      defaultValue="Иван"
      type="text"
    />
  </div>
  <div>
    <span>
      Отчество
    </span>
    <input
      defaultValue="Иванович"
      type="text"
    />
  </div>
  <div>
    <span>
      Баланс
    </span>
    <input
      defaultValue={200}
      type="number"
    />
  </div>
  <button
    className="MobileForm-Save"
    onClick={[Function]}
  >
    Сохранить
  </button>
  <button
    className="MobileForm-Cancel"
    onClick={[Function]}
  >
    Отмена
  </button>
</div>
`;

exports[`работа MobileForm 2`] = `
<button
  className="MobileForm-Create"
  onClick={[Function]}
>
  Создать
</button>
`;

exports[`работа MobileForm 3`] = `
<div
  className="MobileForm"
>
  <div>
    <span>
      Фамилия
    </span>
    <input
      defaultValue="Иванов"
      type="text"
    />
  </div>
  <div>
    <span>
      Имя
    </span>
    <input
      defaultValue="Иван"
      type="text"
    />
  </div>
  <div>
    <span>
      Отчество
    </span>
    <input
      defaultValue="Иванович"
      type="text"
    />
  </div>
  <div>
    <span>
      Баланс
    </span>
    <input
      defaultValue={200}
      type="number"
    />
  </div>
  <button
    className="MobileForm-Save"
    onClick={[Function]}
  >
    Сохранить
  </button>
  <button
    className="MobileForm-Cancel"
    onClick={[Function]}
  >
    Отмена
  </button>
</div>
`;

exports[`работа MobileForm 4`] = `
<button
  className="MobileForm-Create"
  onClick={[Function]}
>
  Создать
</button>
`;

exports[`работа MobileForm 5`] = `
<div
  className="MobileForm"
>
  <div>
    <span>
      Фамилия
    </span>
    <input
      defaultValue="Иванов"
      type="text"
    />
  </div>
  <div>
    <span>
      Имя
    </span>
    <input
      defaultValue="Иван"
      type="text"
    />
  </div>
  <div>
    <span>
      Отчество
    </span>
    <input
      defaultValue="Иванович"
      type="text"
    />
  </div>
  <div>
    <span>
      Баланс
    </span>
    <input
      defaultValue={200}
      type="number"
    />
  </div>
  <button
    className="MobileForm-Save"
    onClick={[Function]}
  >
    Сохранить
  </button>
  <button
    className="MobileForm-Cancel"
    onClick={[Function]}
  >
    Отмена
  </button>
</div>
`;

exports[`работа MobileForm 6`] = `
<button
  className="MobileForm-Create"
  onClick={[Function]}
>
  Создать
</button>
`;
